contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,103 @@
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing  service. Say, try to visit a "host-local" service backend firstly. If no available backend, then try a "zone-local" service backend...
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+```go
+type Service struct {
+  spec:
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // TopologyKeyList is a topologyKey list, which is used to guarantee ability. For example,
+    // `TopologyKeyList: [ "kubernetes.io/hostname", "failure-domain.beta.kubernetes.io/zone" ]`
+    // If failed to match hostname, then it will try to find the endpoints in the same zone
+    // The priority of key in TopologyKeyList is sorted by index and selecting process will stop once a key matches.
+  	TopologyKeyList []string
+}
+```
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname. So, create some annotations for Endpoints which is used toidentify what topology domain the endpoints pod exists, e.g. what host, rack,zone, region etc. In other words, the key of the annotations is the topology key and the value of the annotations is the corresponding value of node's labels.
```

Comment:
We should consider whether we need the `failure-domain...` annotations to be pre-defined, or whether all users of that should switch to a similar mode here.  There's some consistency to be had, I guess, but if topology is arbitrary, we can't predefine every name people might want to use.

Concretely, and extreme position would be to EOL those names, and just let users define whatever keys they want.

I worry that this makes YAMLs less portable, though, so maybe standardizing as many common topology names as we can think of is better?

Reply:
> We should consider whether we need the failure-domain... annotations to be pre-defined

Per your comment below, we have changed the annotations to labels :)

As to pre-define some topology keys or not, there are some well-known labels defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/well_known_labels.go

IMO, topology is arbitrary and users can define whatever they want. That's the value that the proposal respect, isn't it?

Well-known labels defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/well_known_labels.go can be a good example for users and we recommend they switch to a similar mode there, but it's NOT enforced?

Reply:
We simply can't enforce it.  The question is whether we enable some additional portability by defining a set of commonly understood topology labels.  We ave hostname, though it is not explicitly topology.  We have region and zone, though those are called failure-domain (we could rename).  Do we want to define others or are those as far as we can go?

Reply:
> The question is whether we enable some additional portability by defining a set of commonly understood topology labels

Interesting and make sense, I think.

Besides, rack is a common topology key as well though it's not defined yet.